metrics/sources/kubelet/summary_decoders.go

```diff
@@ -0,0 +1,238 @@
+	decodeMemoryStats(nodeMetrics, node.Memory, startTime)
+	decodeNetworkStats(nodeMetrics, node.Network, startTime)
+	decodeFsStats(nodeMetrics, RootFsKey, node.Fs)
+	metrics[NodeKey(node.NodeName)] = nodeMetrics
+
+	for _, container := range node.SystemContainers {
+		key := NodeContainerKey(node.NodeName, container.Name)
+		containerMetrics := decodeContainerStats(labels, &container)
+		containerMetrics.Labels[LabelMetricSetType.Key] = MetricSetTypeSystemContainer
+		metrics[key] = containerMetrics
+	}
+}
+
+func decodePodStats(metrics map[string]*MetricSet, nodeLabels map[string]string, pod *stats.PodStats) {
+	podMetrics := &MetricSet{
+		Labels:         cloneLabels(nodeLabels),
+		MetricValues:   map[string]MetricValue{},
+		LabeledMetrics: []LabeledMetric{},
+	}
+	ref := pod.PodRef
+	podMetrics.Labels[LabelMetricSetType.Key] = MetricSetTypePod
+	podMetrics.Labels[LabelPodId.Key] = ref.UID
+	podMetrics.Labels[LabelPodName.Key] = ref.Name
+	podMetrics.Labels[LabelNamespaceName.Key] = ref.Namespace
+	// Needed for backward compatibility
+	podMetrics.Labels[LabelPodNamespace.Key] = ref.Namespace
+
+	startTime := pod.StartTime.Time
+	decodeUptime(podMetrics, startTime)
+	decodeNetworkStats(podMetrics, pod.Network, startTime)
+	for _, vol := range pod.VolumeStats {
+		decodeFsStats(podMetrics, vol.Name, &vol.FsStats)
+	}
+	metrics[PodKey(ref.Namespace, ref.Name)] = podMetrics
+
+	for _, container := range pod.Containers {
+		key := PodContainerKey(ref.Namespace, ref.Name, container.Name)
+		metrics[key] = decodeContainerStats(podMetrics.Labels, &container)
+	}
+}
+
+func decodeContainerStats(podLabels map[string]string, container *stats.ContainerStats) *MetricSet {
+	containerMetrics := &MetricSet{
+		Labels:         cloneLabels(podLabels),
+		MetricValues:   map[string]MetricValue{},
+		LabeledMetrics: []LabeledMetric{},
+	}
+	containerMetrics.Labels[LabelMetricSetType.Key] = MetricSetTypePodContainer
+	containerMetrics.Labels[LabelContainerName.Key] = container.Name
+
+	startTime := container.StartTime.Time
+	decodeUptime(containerMetrics, startTime)
+	decodeCPUStats(containerMetrics, container.CPU, startTime)
+	decodeMemoryStats(containerMetrics, container.Memory, startTime)
+	decodeFsStats(containerMetrics, RootFsKey, container.Rootfs)
+	decodeFsStats(containerMetrics, LogsKey, container.Logs)
+	decodeUserDefinedMetrics(containerMetrics, container.UserDefinedMetrics)
+
+	return containerMetrics
+}
+
+func decodeUptime(metrics *MetricSet, startTime time.Time) {
+	if startTime.IsZero() {
+		return
+	}
+
+	uptime := uint64(time.Since(startTime).Nanoseconds() / time.Millisecond.Nanoseconds())
+	addIntMetric(metrics, &MetricUptime, &uptime, startTime)
+}
+
+func decodeCPUStats(metrics *MetricSet, cpu *stats.CPUStats, startTime time.Time) {
+	if cpu == nil {
+		return
+	}
+
+	addIntMetric(metrics, &MetricCpuUsage, cpu.UsageCoreNanoSeconds, startTime)
+
+	// Need to translate from nanocores to millicores.
+	if cpu.UsageNanoCores != nil {
```

Comment:
Yes, it is straight from cAdvisor. Currently, it's calculated when the v1 API is translated into the v2 API ([here](https://github.com/google/cadvisor/blob/master/info/v2/conversion.go#L185)), but that will likely change in the future.